pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +182,42 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def upload_one_file(
+        self,
+        file_name,
+        repo_id,
+        type_id,
+        unit_key,
+        unit_metadata=None,
+        override_config=None,
```

Comment:
I'm wondering how much of this do you have a genuine use-case for yet?  If there are some bits which aren't needed currently, I recommend leaving them out instead.

I'm particularly looking at override_config, if you don't have a need to provide this, I think it'd be best to leave it out until it's needed.  It's hard to remove it once it's there, and unfortunately this API requires the caller to know many Pulp implementation details, such as the specific keys understood per override_config.

Actually, can we have some more complete discussion of the use-case first?  I think we need to consider, should we really add a generic upload API or should we rather add some APIs for specific content types we need to handle.

Consider that, if you add a generic upload API, then the users of this method all need to know Pulp implementation details like:

- the valid values for type_id
- what needs to be provided for unit_key
- what can be provided for unit_metadata, override_config

For example, the caller has to know - if I'm uploading an RPM then type_id should be "rpm", but if it's a source RPM then type_id should be "srpm".  I think it's within scope of this library to hide implementation details like that and let the caller interact with Pulp without having to know those kinds of things.

So, where are you planning on using this, should we maybe have an upload API for a specific content type instead?

Reply:
Hm, I intended to upload the file for Maintainance mode in DELIVERY-7160, which is a json file. If just for this use case, I guess most options are not needed here.

Reply:
OK, for that use-case, here is the API I'm thinking of:

```
class FileRepository(...):
  def upload_file(self, filename, relative_url=None):
```

- filename is path to a local file
- relative_url is the path & name which should be used in the remote repo.  It could default to os.path.basename(filename).

unit_key for "iso" units needs name, checksum and size.  checksum and size can be determined during the upload, and name comes from relative_url.  override_config and unit_metadata can be left out until we have some reason to provide them.

The upload method would be on FileRepository class and would be called upload_file, because the API is specific to "file" (or "iso") content and won't be exactly the same in the future for e.g. RPMs.

Reply:
Okay, I got it.